File: 24. Swap Nodes in Pairs/24. Swap Nodes in Pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:

        dummy = ListNode(0, head)
        before = dummy

        while before.next and before.next.next:
            first = before.next
            second = first.next

            before.next = second
            first.next = second.next
            second.next = first

            before = first
        
        return dummy.next

File: 24. Swap Nodes in Pairs/24. Swap Nodes in Pairs.c
struct ListNode
{
    int val;
    struct ListNode *next;
};

struct ListNode *swapPairs(struct ListNode *head)
{
    struct ListNode dummy;
    dummy.val = 0;
    dummy.next = head;

    struct ListNode *before = &dummy;
    struct ListNode *first;
    struct ListNode *second;

    while (before->next && before->next->next)
    {
        first = before->next;
        second = first->next;

        before->next = second;
        first->next = second->next;
        second->next = first;

        before = first;
    }

    return dummy.next;
}